File: frontend/lib/services/firestore_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';

import '../models/conversation_model.dart';
import '../models/journal_model.dart';
import '../models/mood_model.dart';

class FirestoreService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseAuth _auth = FirebaseAuth.instance;

  String? get _uid => _auth.currentUser?.uid;

  DocumentReference get _userDoc {
    if (_uid == null) throw Exception('User not authenticated');
    return _firestore.collection('users').doc(_uid);
  }

  // ============ Conversations ============

  CollectionReference get _conversationsCollection =>
      _userDoc.collection('conversations');

  /// Get all conversations
  Stream<List<ConversationModel>> getConversations() {
    return _conversationsCollection
        .orderBy('updatedAt', descending: true)
        .snapshots()
        .map((snapshot) => snapshot.docs
            .map((doc) => ConversationModel.fromFirestore(doc))
            .toList());
  }

  /// Get a single conversation
  Future<ConversationModel?> getConversation(String id) async {
    final doc = await _conversationsCollection.doc(id).get();
    if (!doc.exists) return null;
    return ConversationModel.fromFirestore(doc);
  }

  /// Create a new conversation
  Future<ConversationModel> createConversation({String? title}) async {
    final now = DateTime.now();
    final docRef = _conversationsCollection.doc();
    
    final conversation = ConversationModel(
      id: docRef.id,
      title: title ?? 'New Conversation',
      messageCount: 0,
      createdAt: now,
      updatedAt: now,
    );
    
    await docRef.set(conversation.toFirestore());
    return conversation;
  }

  /// Update conversation title
  Future<void> updateConversationTitle(String id, String title) async {
    await _conversationsCollection.doc(id).update({
      'title': title,
      'updatedAt': FieldValue.serverTimestamp(),
    });
  }

  /// Delete a conversation
  Future<void> deleteConversation(String id) async {
    // Delete all messages first
    final messages = await _conversationsCollection
        .doc(id)
        .collection('messages')
        .get();
    
    for (final doc in messages.docs) {
      await doc.reference.delete();
    }
    
    await _conversationsCollection.doc(id).delete();
  }

  // ============ Messages ============

  CollectionReference _messagesCollection(String conversationId) =>
      _conversationsCollection.doc(conversationId).collection('messages');

  /// Get messages for a conversation
  Stream<List<MessageModel>> getMessages(String conversationId) {
    return _messagesCollection(conversationId)
        .orderBy('createdAt')
        .snapshots()
        .map((snapshot) =>
            snapshot.docs.map((doc) => MessageModel.fromFirestore(doc)).toList());
  }

  /// Add a message to a conversation
  Future<MessageModel> addMessage({
    required String conversationId,
    required String role,
    required String content,
  }) async {
    final docRef = _messagesCollection(conversationId).doc();
    final now = DateTime.now();
    
    final message = MessageModel(
      id: docRef.id,
      role: role,
      content: content,
      createdAt: now,
    );
    
    await docRef.set(message.toFirestore());
    
    // Update conversation
    await _conversationsCollection.doc(conversationId).update({
      'messageCount': FieldValue.increment(1),
      'updatedAt': FieldValue.serverTimestamp(),
    });
    
    return message;
  }

  // ============ Journals ============

  CollectionReference get _journalsCollection => _userDoc.collection('journals');

  /// Get all journals
  Stream<List<JournalModel>> getJournals() {
    return _journalsCollection
        .orderBy('createdAt', descending: true)
        .snapshots()
        .map((snapshot) =>
            snapshot.docs.map((doc) => JournalModel.fromFirestore(doc)).toList());
  }

  /// Get a single journal
  Future<JournalModel?> getJournal(String id) async {
    final doc = await _journalsCollection.doc(id).get();
    if (!doc.exists) return null;
    return JournalModel.fromFirestore(doc);
  }

  /// Create a new journal entry
  Future<JournalModel> createJournal({
    required String content,
    String? moodId,
  }) async {
    final now = DateTime.now();
    final docRef = _journalsCollection.doc();
    
    final journal = JournalModel(
      id: docRef.id,
      content: content,
      moodId: moodId,
      createdAt: now,
      updatedAt: now,
    );
    
    await docRef.set(journal.toFirestore());
    return journal;
  }

  /// Update a journal entry
  Future<void> updateJournal({
    required String id,
    String? content,
    String? moodId,
    String? aiInsight,
  }) async {
    final updates = <String, dynamic>{
      'updatedAt': FieldValue.serverTimestamp(),
    };
    
    if (content != null) updates['content'] = content;
    if (moodId != null) updates['moodId'] = moodId;
    if (aiInsight != null) updates['aiInsight'] = aiInsight;
    
    await _journalsCollection.doc(id).update(updates);
  }

  /// Delete a journal entry
  Future<void> deleteJournal(String id) async {
    await _journalsCollection.doc(id).delete();
  }

  // ============ Moods ============

  CollectionReference get _moodsCollection => _userDoc.collection('moods');

  /// Get all moods
  Stream<List<MoodModel>> getMoods() {
    return _moodsCollection
        .orderBy('date', descending: true)
        .snapshots()
        .map((snapshot) =>
            snapshot.docs.map((doc) => MoodModel.fromFirestore(doc)).toList());
  }

  /// Get moods for a date range
  Future<List<MoodModel>> getMoodsInRange(DateTime start, DateTime end) async {
    final snapshot = await _moodsCollection
        .where('date', isGreaterThanOrEqualTo: Timestamp.fromDate(start))
        .where('date', isLessThanOrEqualTo: Timestamp.fromDate(end))
        .orderBy('date')
        .get();
    
    return snapshot.docs.map((doc) => MoodModel.fromFirestore(doc)).toList();
  }

  /// Get today's mood
  Future<MoodModel?> getTodaysMood() async {
    final now = DateTime.now();
    final startOfDay = DateTime(now.year, now.month, now.day);
    final endOfDay = startOfDay.add(const Duration(days: 1));
    
    final snapshot = await _moodsCollection
        .where('date', isGreaterThanOrEqualTo: Timestamp.fromDate(startOfDay))
        .where('date', isLessThan: Timestamp.fromDate(endOfDay))
        .limit(1)
        .get();
    
    if (snapshot.docs.isEmpty) return null;
    return MoodModel.fromFirestore(snapshot.docs.first);
  }

  /// Create or update today's mood
  Future<MoodModel> saveMood({
    required int score,
    required String emoji,
    String? note,
  }) async {
    final existingMood = await getTodaysMood();
    
    if (existingMood != null) {
      // Update existing mood
      await _moodsCollection.doc(existingMood.id).update({
        'score': score,
        'emoji': emoji,
        'note': note,
      });
      return existingMood.copyWith(score: score, emoji: emoji, note: note);
    } else {
      // Create new mood
      final docRef = _moodsCollection.doc();
      final mood = MoodModel(
        id: docRef.id,
        date: DateTime.now(),
        score: score,
        emoji: emoji,
        note: note,
      );
      
      await docRef.set(mood.toFirestore());
      return mood;
    }
  }

  /// Delete a mood entry
  Future<void> deleteMood(String id) async {
    await _moodsCollection.doc(id).delete();
  }
}

File: backend/app/services/firestore_service.py
from datetime import datetime
from typing import List, Optional
from google.cloud.firestore_v1 import FieldFilter
import uuid

from ..dependencies import get_firestore_client


class FirestoreService:
    """Service for Firestore operations."""
    
    def __init__(self):
        self._db = get_firestore_client()
    
    def _user_ref(self, uid: str):
        """Get reference to user document."""
        return self._db.collection("users").document(uid)
    
    # ============ Conversations ============
    
    async def get_conversations(self, uid: str) -> List[dict]:
        """Get all conversations for a user."""
        conversations_ref = self._user_ref(uid).collection("conversations")
        docs = conversations_ref.order_by("updatedAt", direction="DESCENDING").stream()
        
        return [
            {
                "id": doc.id,
                **doc.to_dict(),
            }
            for doc in docs
        ]
    
    async def get_conversation(self, uid: str, conversation_id: str) -> Optional[dict]:
        """Get a single conversation."""
        doc = self._user_ref(uid).collection("conversations").document(conversation_id).get()
        if not doc.exists:
            return None
        return {"id": doc.id, **doc.to_dict()}
    
    async def create_conversation(self, uid: str, title: str = "New Conversation") -> dict:
        """Create a new conversation."""
        conversations_ref = self._user_ref(uid).collection("conversations")
        doc_ref = conversations_ref.document()
        
        now = datetime.utcnow()
        data = {
            "title": title,
            "messageCount": 0,
            "createdAt": now,
            "updatedAt": now,
        }
        
        doc_ref.set(data)
        return {"id": doc_ref.id, **data}
    
    async def update_conversation(
        self,
        uid: str,
        conversation_id: str,
        title: Optional[str] = None,
        increment_messages: bool = False,
    ):
        """Update a conversation."""
        doc_ref = self._user_ref(uid).collection("conversations").document(conversation_id)
        
        updates = {"updatedAt": datetime.utcnow()}
        
        if title:
            updates["title"] = title
        
        doc_ref.update(updates)
        
        if increment_messages:
            from google.cloud.firestore_v1 import Increment
            doc_ref.update({"messageCount": Increment(1)})
    
    # ============ Messages ============
    
    async def get_messages(
        self,
        uid: str,
        conversation_id: str,
        limit: int = 50,
    ) -> List[dict]:
        """Get messages for a conversation."""
        messages_ref = (
            self._user_ref(uid)
            .collection("conversations")
            .document(conversation_id)
            .collection("messages")
        )
        
        docs = messages_ref.order_by("createdAt").limit(limit).stream()
        
        return [
            {
                "id": doc.id,
                **doc.to_dict(),
            }
            for doc in docs
        ]
    
    async def add_message(
        self,
        uid: str,
        conversation_id: str,
        role: str,
        content: str,
    ) -> dict:
        """Add a message to a conversation."""
        messages_ref = (
            self._user_ref(uid)
            .collection("conversations")
            .document(conversation_id)
            .collection("messages")
        )
        
        doc_ref = messages_ref.document()
        now = datetime.utcnow()
        
        data = {
            "role": role,
            "content": content,
            "createdAt": now,
        }
        
        doc_ref.set(data)
        
        # Update conversation
        await self.update_conversation(uid, conversation_id, increment_messages=True)
        
        return {"id": doc_ref.id, **data}
    
    # ============ Moods ============
    
    async def get_moods_in_range(
        self,
        uid: str,
        start_date: datetime,
        end_date: datetime,
    ) -> List[dict]:
        """Get mood entries within a date range."""
        moods_ref = self._user_ref(uid).collection("moods")
        
        docs = (
            moods_ref
            .where(filter=FieldFilter("date", ">=", start_date))
            .where(filter=FieldFilter("date", "<=", end_date))
            .order_by("date")
            .stream()
        )
        
        return [
            {
                "id": doc.id,
                **doc.to_dict(),
            }
            for doc in docs
        ]


# Singleton instance
_firestore_service = None


def get_firestore_service() -> FirestoreService:
    """Get or create FirestoreService instance."""
    global _firestore_service
    if _firestore_service is None:
        _firestore_service = FirestoreService()
    return _firestore_service
